neqo-http3/src/stream_type_reader.rs

```diff
@@ -6,57 +6,120 @@
             stream_id,
         }
     }
 
-    pub fn get_type(&mut self, conn: &mut Connection) -> Option<Http3StreamType> {
-        // On any error we will only close this stream!
-        loop {
-            if let NewStreamTypeReader::Read {
-                ref mut reader,
-                stream_id,
-            } = self
-            {
+    fn read(&mut self, conn: &mut Connection) -> Res<(Option<u64>, bool)> {
+        match self {
+            NewStreamTypeReader::ReadType {
+                reader, stream_id, ..
+            }
+            | NewStreamTypeReader::ReadSessionId { reader, stream_id } => loop {
                 let to_read = reader.min_remaining();
                 let mut buf = vec![0; to_read];
-                match conn.stream_recv(*stream_id, &mut buf[..]) {
-                    Ok((0, false)) => {
-                        return None;
-                    }
-                    Ok((amount, false)) => {
-                        if let Some(res) = reader.consume(&mut Decoder::from(&buf[..amount])) {
-                            *self = NewStreamTypeReader::Done;
-                            return Some(res.into());
+                match conn.stream_recv(*stream_id, &mut buf[..])? {
+                    (0, f) => break Ok((None, f)),
+                    (amount, f) => {
+                        let res = reader.consume(&mut Decoder::from(&buf[..amount]));
+                        if res.is_some() || f {
+                            break Ok((res, f));
                         }
                     }
-                    Ok((_, true)) | Err(_) => {
+                }
+            },
+            _ => Ok((None, false)),
+        }
+    }
+
+    pub fn get_type(&mut self, conn: &mut Connection) -> Res<Option<NewStreamType>> {
+        loop {
+            let (output, fin) = self.read(conn)?;
+            if output.is_none() {
+                if fin {
+                    *self = NewStreamTypeReader::Done;
+                    break Err(Error::HttpStreamCreation);
+                }
+                break Ok(None);
+            }
+            let output = output.unwrap();
+            qtrace!("Decoded uint {}", output);
+            match self {
+                NewStreamTypeReader::ReadType {
+                    push_stream_allowed,
+                    stream_id,
+                    ..
+                } => {
+                    let res = NewStreamType::decode_stream_type(output, *push_stream_allowed);
+                    if res.is_err() {
+                        *self = NewStreamTypeReader::Done;
+                        return res;
+                    }
+                    let res = res.unwrap();
+                    if fin {
+                        *self = NewStreamTypeReader::Done;
+                        return self.map_stream_fin(res);
+                    }
+                    qtrace!("Decoded stream type {:?}", res);
+                    if res.is_some() {
```

Comment:
It's not clear from context here what `res` is.  Naming this variable better (and the decode_stream_type function) would make this easier to understand.  The function returns two things: a concrete stream type in the case where the stream header is complete, or `None` if additional information is required from the header to produce a final type.  You could rename it `final_stream_type`.

Some commentary might help as well.  That is, "if the stream type isn't final, we need to read an extra identifier from the stream to know what the stream is for".